week2/part4/tensor.hpp

```diff
@@ -44,59 +44,251 @@ class Tensor
     [[nodiscard]] size_t rank() const;
 
     // Returns the shape of the tensor.
-    [[nodiscard]] std::vector< size_t > shape() const;
+    [[nodiscard]] std::vector<size_t> shape() const;
 
     // Returns the number of elements of this tensor.
     [[nodiscard]] size_t numElements() const;
 
     // Element access function
-    const ComponentType&
-    operator()(const std::vector< size_t >& idx) const;
+    const ComponentType &
+    operator()(const std::vector<size_t> &idx) const;
 
     // Element mutation function
-    ComponentType&
-    operator()(const std::vector< size_t >& idx);
+    ComponentType &
+    operator()(const std::vector<size_t> &idx);
 
-private:
-    // TODO: Probably you need some members here...
+    // Getters
+    const std::vector<ComponentType> &getData() const { return data_; }
+    const std::vector<size_t> &getShape() const { return shape_; }
 
+    // Setters
+    void setData(const std::vector<ComponentType> &data) { data_ = data; }
+    void setShape(const std::vector<size_t> &shape) { shape_ = shape; }
+
+private:
+    std::vector<size_t> shape_;       // Represents the shape of the tensor
+    std::vector<ComponentType> data_; // Represents the tensor's elements
 };
 
-// TODO: Implement all methods of the Tensor class template.
+template <Arithmetic ComponentType>
+Tensor<ComponentType>::Tensor() : shape_(1), data_(numElements(), 0)
+{
+    // Set shape to an empty vector, indicating rank 0
+    // Set data_ with a single, zero-initialized entry
+}
```

Comment:
FYI this implementation is based on my understanding of the problem statement (extract below): 

> Constructors. Implement a primary constructor that receives the shape of the tensor. The shape defines
the number of elements in each dimension. An empty shape vector results in a scalar, a shape of size 1
is equivalent to a vector with shape[0] elements. A shape of size 2 results in a tensor that resembles a
matrix. The number of dimensions (or size of the shape-vector) is also referred to as rank. The default
constructor shall create a tensor with rank 0, containing a single, zero-initialized entry. Use std::fill
to fill up the memory with the fillValue if specified.